Catwalk Tiles REVAMP + floorbot buff

OrionTheFox opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Requested to bring this upstream, so here we are now. The floorbot stuff is new though.

Adds new catwalk tiles that blend better with other flooring
![image](https://user-images.githubusercontent.com/76465278/142985371-670b6cd2-9c93-41a2-b595-52d1d6eb2c1f.png)

Also does some improvements to floorbots, specifically:
-Auto-place catwalks in maints (now when they wander in there they'll actually be helpful!)
-Actually patch holes (they used to just lay a tile over space and call it good; crowbarring it up would show a lack of plating)
-Repairing floor tiles actually REPAIRS it, leaving the decals intact

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
More usable catwalks, rather than what we currently had.
<hr>

<details><summary>A BOMB HAS GONE OFF IN THE BRIG</summary>

![image](https://user-images.githubusercontent.com/76465278/142985790-32208dae-0aa1-4636-ad4e-9d99a713e47a.png)</details>
<details><summary>Summon the floor bots--- HEY!</summary>

![image](https://user-images.githubusercontent.com/76465278/142985851-d0506eee-e49f-426b-92b8-eb3d95a635e9.png)
</details>
<details><summary>Build the floor bots, and repair the breach!</summary>

![image](https://user-images.githubusercontent.com/76465278/142985905-81e2812c-b849-4c06-80d3-7bef0172132d.png)
</details>
<details><summary>The improved Floorbot Repair Function from SS13</summary>

![image](https://user-images.githubusercontent.com/76465278/142985946-f0c3cb96-5727-4d58-ae80-0175bce8442b.png)
</details>
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Added new reskins of the Catwalk Tiles, which will blend better with other types of floor! These can be selected via a radial on the base catwalk tile, crafted with iron rods.
expansion: Floorbots will automatically place catwalks when they're in a maintenance area, and will repair damaged/burnt tiles rather than fully replacing them.
fix: Floorbots will actually place a plating when repairing a breach with tiles, rather than haphazardly laying a tile that has space below it.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Commit:
catwalks hnngh

Commit:
floorbots too :)

Commit:
conflict moment

OrionTheFox commented:
pinging @tralezab as you wanted the tiles here; the floorbot stuff is still up for debate tho

Arctaisia commented:
👀

LemonInTheDark commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -24,11 +25,11 @@
 	. = ..()
-	var/static/image/catwalk_overlay
+	var/image/catwalk_overlay	///Per-tile instead of static, so we can have multiple different types of catwalk
 	if(isnull(catwalk_overlay))
 		catwalk_overlay = new()
 		catwalk_overlay.icon = icon
-		catwalk_overlay.icon_state = "catwalk_above"
+		catwalk_overlay.icon_state = "[above_state]"
```
This does not spark joy
Please cache this

LemonInTheDark replied:
Each time update_overlays is called, what we'd have now would need to generate a new catwalk overlay. This is a waste of raw time because it's the same overlay over and over again, and it's also a waste of memory because we may have say 10000 platings all with unique images.

Static vars inside a proc like that one persist through calls, so there'd only ever be 1 catwalk overlay.
What you've got now changes that slightly, since now you need a new catwalk overlay per above state.
I'd like you to cache this somehow. Maybe a global list of above_state -> catwalk overlay, then in this proc you check if the list has the overlay you want, if it does you use it and if not you generate one and add it to the list

OrionTheFox replied:
```
GLOBAL_LIST_INIT(catwalk_overlay_masterlist, list())	//Stores all the above_states for the different types of catwalk

/turf/open/floor/catwalk_floor/update_overlays()
	. = ..()
	if(covered)
		if(above_state in GLOB.catwalk_overlay_masterlist)
			. += GLOB.catwalk_overlay_masterlist[above_state]
		else
			var/image/catwalk_overlay = new()
			catwalk_overlay.icon = icon
			catwalk_overlay.icon_state = above_state
			catwalk_overlay.plane = GAME_PLANE
			catwalk_overlay.layer = CATWALK_LAYER
			GLOB.catwalk_overlay_masterlist[above_state] = catwalk_overlay
			. += catwalk_overlay```
new code look good?

OrionTheFox replied:
wow i need to learn formatting better
@LemonInTheDark

OrionTheFox replied:
i'm gonna actually mark this as resolved, if the new code has an issue re-review that section pls

LemonInTheDark commented on code/modules/mob/living/simple_animal/bot/floorbot.dm:
```diff
@@ -368,12 +370,20 @@
+				F.broken = FALSE
+				F.burnt = FALSE
+				F.icon_state = initial(F.icon_state)
+			else if(findtext(is_this_maints.name, "maint"))	//place catwalk if it's plating and we're in maints
```
This is not an acceptable way to detect maintenance. I think there's flags you can use?

timothymtorres replied:
I code dived and the only two area helper procs are `isarea()` and `get_area()` both of which cannot be used to check if a turf is maint.

Then I looked at the flags as an option:

https://github.com/tgstation/tgstation/blob/a3a174f031ecd81773f55bffef1f8f152fddada2/code/game/area/space_station_13_areas.dm#L168-L173

https://github.com/tgstation/tgstation/blob/11302feef45d2bc0fc717ee426114c2c7e52caeb/code/_globalvars/bitfields.dm#L52-L70

It doesn't appear there is a maintenance flag set anywhere for the codebase.  Looks like a new one will need to be made.  I wonder if others used similar hacks like this to check if a turf was maintenance...

OrionTheFox replied:
this was the method used for floorbots' `is_hull_breach()` proc, i assumed it'd be fine

`/mob/living/simple_animal/bot/floorbot/proc/is_hull_breach(turf/t) //Ignore space tiles not considered part of a structure, also ignores shuttle docking areas.`
`>>>var/area/t_area = get_area(t)`
`>>>if(t_area && (t_area.name == "Space" || findtext(t_area.name, "huttle")))`
`(more code past here`

OrionTheFox replied:
as for the multi-z lighting thing, it most likely doesnt fix nor damage that; these are all just reskins of the original catwalk

LemonInTheDark replied:
> I code dived and the only two area helper procs are `isarea()` and `get_area()` both of which cannot be used to check if a turf is maint.
> 
> Then I looked at the flags as an option:

Sorry should have been more clear. They should add a bitflag that's "is maintenance" rather then searching the name, because names can change and that's not a stable way to check

I have no doubt boomer code did this too, but I dream of a better tomorrow

OrionTheFox replied:
I can tryyyy to fix this? Maybe? Lemme look into it.

OrionTheFox replied:
added, called "IS_MAINTINENCE_AREA"

OrionTheFox replied:
marking this as resolved, if i did something wrong please re-review hnnnnhbhhgbg

LemonInTheDark commented on code/modules/mob/living/simple_animal/bot/floorbot.dm:
```diff
@@ -368,12 +370,20 @@
+			else if(F.broken || F.burnt)	//repair the tile and reset it to be undamaged (rather than replacing it)
+				F.broken = FALSE
+				F.burnt = FALSE
+				F.icon_state = initial(F.icon_state)
```
this smells like it might break, you sure inital is the right way to do this?

OrionTheFox replied:
already tested it, `initial(F.icon_state)` checks the tile's icon state when it was first loaded; unfortunately there's no proc for UNbreaking/UNburning a tile, so this was the simplest way to do it

OrionTheFox replied:
but yes, after pretty thorough testing, it didnt break. and if it does in the future, that's most likely an issue of the placed tile itself 🦀

Dorsisdwarf commented:
Is it too harsh to say that I kinda feel like if these are for blending catwalks with other floors they should do a smart connection or something? The sprite is non-tiling and doesnt blend with, well, catwalks. It just provides "catwalk that looks like an ugly floor tile" as an option

NanoCats commented:
As a floorbot enjoyer, I will really enjoy this.


I've wanted to set loose a herd of floorbots to catwalk-tile the entire station, but currently floorbots themselves are broken (and have been for a while #62880 ) 

Also they don't properly show wires/pipes for some reason.

EricZilla commented:
I really like those blended catwalks, makes me hope mappers will use them in areas where you need access to wires

OrionTheFox commented:
> Is it too harsh to say that I kinda feel like if these are for blending catwalks with other floors they should do a smart connection or something? The sprite is non-tiling and doesnt blend with, well, catwalks themself. It just provides "catwalk that looks like an ugly floor tile" as an option

I actually contemplated smoothing them together heavily, but came across the big issue of "you're supposed to unscrew the top" and realized that unscrewing a random chunk of a seemingly smooth catwalk was more ugly than this alternative

OrionTheFox commented:
> As a floorbot enjoyer, I will really enjoy this.
> 
> I've wanted to set loose a herd of floorbots to catwalk-tile the entire station, but currently floorbots themselves are broken (and have been for a while #62880 )
> 
> Also they don't properly show wires/pipes for some reason.

I can say the floorbots were at least TRYING, I dunno if they're fully fixed or not. Took a bit of force as an AI to get them going.
The catwalks show wires/pipes now that the fix PR is merged 😎 that's what this PR was waiting on

NanoCats commented:
> > As a floorbot enjoyer, I will really enjoy this.
> > I've wanted to set loose a herd of floorbots to catwalk-tile the entire station, but currently floorbots themselves are broken (and have been for a while #62880 )
> > Also they don't properly show wires/pipes for some reason.
> 
> I can say the floorbots were at least TRYING, I dunno if they're fully fixed or not. Took a bit of force as an AI to get them going. The catwalks show wires/pipes now that the fix PR is merged 😎 that's what this PR was waiting on

They'd _better_  ![image](https://user-images.githubusercontent.com/14065903/142344510-3e8dd70b-7c86-4f4b-b3a7-90c132ca0c51.png)
\s

datorangebottle commented:
A man has fallen into the river in space station 13!

Commit:
i think this is right

Commit:
IS_MAINTINENCE_AREA bitflag

Commit:
i apparently cant spell

OrionTheFox commented on code/__DEFINES/flags.dm:
```diff
@@ -120,6 +120,8 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 /// If engravings are persistent in this area
 #define PERSISTENT_ENGRAVINGS (1<<15)
+///If this area qualifies as Maintenence (or similar)
+#define IS_MAINTENENCE_AREA (1<<16)
```
```suggestion
#define IS_MAINTENANCE_AREA (1<<16)
```
jesus christ i just cannot spell

OrionTheFox commented on code/_globalvars/bitfields.dm:
```diff
@@ -67,6 +67,7 @@ DEFINE_BITFIELD(area_flags, list(
 	"NO_ALERTS" = NO_ALERTS,
 	"AREA_USES_STARLIGHT" = AREA_USES_STARLIGHT,
 	"PERSISTENT_ENGRAVINGS" = PERSISTENT_ENGRAVINGS,
+	"IS_MAINTENENCE_AREA" = IS_MAINTENENCE_AREA,
```
```suggestion
	"IS_MAINTENANCE_AREA" = IS_MAINTENANCE_AREA,
```

OrionTheFox commented on code/game/area/space_station_13_areas.dm:
```diff
@@ -168,7 +168,7 @@ NOTE: there are two lists of areas in the end of this file: centcom and station
 	name = "Generic Maintenance"
 	ambience_index = AMBIENCE_MAINT
-	area_flags = BLOBS_ALLOWED | UNIQUE_AREA | CULT_PERMITTED | PERSISTENT_ENGRAVINGS
+	area_flags = BLOBS_ALLOWED | UNIQUE_AREA | CULT_PERMITTED | PERSISTENT_ENGRAVINGS | IS_MAINTENENCE_AREA
```
```suggestion
	area_flags = BLOBS_ALLOWED | UNIQUE_AREA | CULT_PERMITTED | PERSISTENT_ENGRAVINGS | IS_MAINTENANCE_AREA
```

OrionTheFox commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -5,9 +5,9 @@
  * unless you want to!
  */
-/turf/open/floor/catwalk_floor
+/turf/open/floor/catwalk_floor	//the base type, meant to look like a maintenence panel
```
```suggestion
/turf/open/floor/catwalk_floor	//the base type, meant to look like a maintenance panel
```

OrionTheFox commented on code/modules/mob/living/simple_animal/bot/floorbot.dm:
```diff
@@ -368,12 +370,20 @@
+				F.broken = FALSE
+				F.burnt = FALSE
+				F.icon_state = initial(F.icon_state)
+			else if(is_this_maints.area_flags == IS_MAINTENENCE_AREA)	//place catwalk if it's plating and we're in maints
```
```suggestion
			else if(is_this_maints.area_flags == IS_MAINTENANCE_AREA)	//place catwalk if it's plating and we're in maints
```

OrionTheFox commented on code/game/objects/items/stacks/tiles/tile_types.dm:
```diff
@@ -1113,18 +1113,52 @@
 
-/obj/item/stack/tile/catwalk_tile
+//Catwalk Tiles
+/obj/item/stack/tile/catwalk_tile //This is our base type, sprited to look maintenence-styled
```
```suggestion
/obj/item/stack/tile/catwalk_tile //This is our base type, sprited to look maintenance-styled
```

OrionTheFox reviewed (commented):
its 4am and i dont know how to spell maintenance apparently

Commit:
Apply suggestions from code review because I cannae spell

LemonInTheDark commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -16,24 +16,28 @@
-		catwalk_overlay.plane = GAME_PLANE
-		catwalk_overlay.layer = CATWALK_LAYER
-		catwalk_overlay = catwalk_overlay.appearance
 	if(covered)
```
`if(covered)` should be replaced with `if(!covered) ...(next line) return` so we don't indent too deep

LemonInTheDark replied:
This gives a good explanation as to why https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#use-early-returns

OrionTheFox replied:
I felt sick coding that line, this is why, thanks for calling me out LMAO

LemonInTheDark commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -16,24 +16,28 @@
-		catwalk_overlay = catwalk_overlay.appearance
 	if(covered)
-		. += catwalk_overlay
+		if(above_state in GLOB.catwalk_overlay_masterlist)
+			. += GLOB.catwalk_overlay_masterlist[above_state]
+		else
```
may be a bit cleaner to do 
```
if(!GLOB.catwalk_overlay_masterlist[above_state])
    //Generate new overlay
    GLOB.catwalk_overlay_masterlist[above_state] = catwalk_overlay
. += GLOB.catwalk_overlay_masterlist[above_state]
```
rather then adding to . twice. Removes the need for the else too

LemonInTheDark commented on code/game/area/space_station_13_areas.dm:
```diff
@@ -168,7 +168,7 @@ NOTE: there are two lists of areas in the end of this file: centcom and station
 /area/maintenance
 	name = "Generic Maintenance"
```
Are there any other areas with maint in their name that could use the flag?

OrionTheFox replied:
to my surprise, this was /ACTUALLY/ done consistently. anywhere that should be maintinence is a subtype of this area; the ONE exception is the AI sat, which I'm gonna wait on a specific 'do that' before flagging - it's such an edgecase that I think the whole AI sat would probably just have a unique flag at that point anyways

the one flaw i realized with this is that Radstorm Shelters are ALSO subtypes of maintinence, so they'll get catwalked if they lose tiles. Not a big flaw, but if radstorms ever come back in full force it might be worth a flag specifically for rad-proofing instead of just using maintinence for it

LemonInTheDark replied:
Make sure you add the flag to any subtypes that override area_flags, xenobio maint comes to mind.

I feel a bit silly proposing this, but if we just want all the subtypes of /area/maintenance, why not just do `istype(area, /area/maintenance)`? or is having more granular control useful you think

OrionTheFox replied:
well
tbh
I didnt know ALL maints were properly subtypes of maints. Figured someone'd messed that up at some point, so I searched the area's name for "maint";; but surprisingly they havent messed it up so
this bitflag stuff might be a bit pointless thinking about it, but idk

LemonInTheDark replied:
you wanna uh, just use the typepath?
Seems close enough to "is this maintenance" to me, genuinely kinda feel bad I didn't realize this earlier

OrionTheFox replied:
yeah i think I'm gonna bail on the bitflag and just check if its area/maintenance

if we need a bitflag, it should be for something more specific, like 'radproof', not just 'maints'; maints really SHOULD just be 'subtype of maints'

LemonInTheDark commented on code/modules/mob/living/simple_animal/bot/floorbot.dm:
```diff
@@ -368,12 +370,20 @@
+				F.broken = FALSE
+				F.burnt = FALSE
+				F.icon_state = initial(F.icon_state)
+			else if(is_this_maints.area_flags == IS_MAINTENANCE_AREA)	//place catwalk if it's plating and we're in maints
```
`area_flags` is a bitflield, which means it's effectively one big number, and we're individually flipping its bits to store true/false stuff, so there's a bit that's for maintenance, if it's flipped on we're maint, if it's not we're not, etc.

You can't check that with just ==, since it will test if the whole flags number is equal to 1 set bit. That'll fail, since we store more then just that in `area_flags`.

You'll want to use binary operators. & will check if two numbers share any bitflags, so if you & `area_flags` and `IS_MAINTENANCE_AREA` together you'll be able to see if `area_flags` has `IS_MAINTENANCE_AREA` set.

You'll want to read this wiki page https://tgstation13.org/wiki/Binary_flags
It gives a really pretty decent rundown on the idea. Only thing to note is 1<<1 is setting the second bit to 1, 1<<5 sets the fifth. That's not discussed in the wiki page I think because it didn't used to work? They use the raw numbers instead, so 1, 2, 4, 8, 16, 32, etc.

If anythings confusing feel free to bug me on discord or whatever

OrionTheFox replied:
Yeah, I think I get it. Was wondering what the 1 < # meant but that explains it pretty well.
I _think_ I wrote it correctly, call me out if I did it wrong but I at least understand what it's meant to be doing now

LemonInTheDark replied:
Ya did good

LemonInTheDark commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -16,24 +16,28 @@
 	. = ..()
 	update_icon(UPDATE_OVERLAYS)
 
+GLOBAL_LIST_INIT(catwalk_overlay_masterlist, list())	//Stores all the above_states for the different types of catwalk
```
Stores all the different types of overlays for the catwalk, right? Not just all the icon states.

You may want `GLOBAL_LIST_EMPTY` here too instead of `GLOBAL_LIST_INIT`, it gets rid of the need for that second argument

OrionTheFox replied:
yeah, it stores the icon_state paired with the overlay variables, so all you need to do is select it and it'll be on the proper layer and everything
at least, thats the intention. it seems to work that way too, so i HOPEFULLY did it right

LemonInTheDark commented on code/game/objects/items/stacks/tiles/tile_types.dm:
```diff
@@ -1113,18 +1113,52 @@
+	icon_state = "titanium_catwalk"
+	turf_type = /turf/open/floor/catwalk_floor/titanium
+
+/obj/item/stack/tile/catwalk_tile/iron_smooth //this is the greenish one
```
Why call it smooth if it's greenish?

OrionTheFox replied:
the tile its based off of is iron/smooth or something similar, I figured it'd be better to keep it similarly named

LemonInTheDark reviewed (changes requested):
Very good, thank you.
Some more review, getting into bitflags, feel free to bug me if it confuses you

OrionTheFox commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -16,24 +16,28 @@
 	. = ..()
 	update_icon(UPDATE_OVERLAYS)
 
+GLOBAL_LIST_INIT(catwalk_overlay_masterlist, list())	//Stores all the above_states for the different types of catwalk
```
```suggestion
GLOBAL_LIST_EMPTY(catwalk_overlay_masterlist)	//Stores all the above_states for the different types of catwalk
```

OrionTheFox commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -16,24 +16,28 @@
-		catwalk_overlay.plane = GAME_PLANE
-		catwalk_overlay.layer = CATWALK_LAYER
-		catwalk_overlay = catwalk_overlay.appearance
 	if(covered)
```
```suggestion
	if(!covered)
		return
```

OrionTheFox commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -16,24 +16,28 @@
-		catwalk_overlay = catwalk_overlay.appearance
 	if(covered)
-		. += catwalk_overlay
+		if(above_state in GLOB.catwalk_overlay_masterlist)
+			. += GLOB.catwalk_overlay_masterlist[above_state]
+		else
+			var/image/catwalk_overlay = new()
+			catwalk_overlay.icon = icon
+			catwalk_overlay.icon_state = above_state
+			catwalk_overlay.plane = GAME_PLANE
+			catwalk_overlay.layer = CATWALK_LAYER
+			GLOB.catwalk_overlay_masterlist[above_state] = catwalk_overlay
+			. += catwalk_overlay
```
```suggestion
		if(!GLOB.catwalk_overlay_masterlist[above_state])
			//Generate a new overlay and add it to the global list
			var/image/catwalk_overlay = new()
			catwalk_overlay.icon = icon
			catwalk_overlay.icon_state = above_state
			catwalk_overlay.plane = GAME_PLANE
			catwalk_overlay.layer = CATWALK_LAYER
			GLOB.catwalk_overlay_masterlist[above_state] = catwalk_overlay
		. += GLOB.catwalk_overlay_masterlist[above_state]
```

OrionTheFox commented on code/modules/mob/living/simple_animal/bot/floorbot.dm:
```diff
@@ -368,12 +370,20 @@
+				F.broken = FALSE
+				F.burnt = FALSE
+				F.icon_state = initial(F.icon_state)
+			else if(is_this_maints.area_flags == IS_MAINTENANCE_AREA)	//place catwalk if it's plating and we're in maints
```
```suggestion
			else if(is_this_maints.area_flags & IS_MAINTENANCE_AREA)	//place catwalk if it's plating and we're in maints
```

OrionTheFox reviewed (commented):
hnngh I think these'll be right

Commit:
Apply suggestions from code review (hopefully correct)

OrionTheFox commented:
I swear every time I make a change it automatically requests a review from a new person wtf

OrionTheFox commented:
web edit moment, broke something, i'll fix this

LemonInTheDark commented:
make sure to test it locally, you or I could easily be missing something

Commit:
indent + github webedit moment

OrionTheFox commented:
the test was planned, just was fixing the error and a few more checks before hopefully the final test

OrionTheFox commented:
alright @LemonInTheDark it works beautifully

floorbots are hectic and untrustworthy, but that's not my fault. My part of their code works just how it should, repairs and catwalks in maints.

OrionTheFox commented:
just gotta remove this
one comment...

Commit:
hnngh it worked

Commit:
I still cant spell

OrionTheFox commented:
> 
> floorbots are hectic and untrustworthy, but that's not my fault. My part of their code works just how it should, repairs and catwalks in maints.

as I was saying
![image](https://user-images.githubusercontent.com/76465278/143538290-2630c0f8-9ab1-4aa5-a623-02f34b1dbdcf.png)
bot pathing is something of its own

OrionTheFox commented on code/__DEFINES/flags.dm:
```diff
@@ -120,6 +120,8 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 #define AREA_USES_STARLIGHT (1<<14)
 /// If engravings are persistent in this area
 #define PERSISTENT_ENGRAVINGS (1<<15)
+///If this area qualifies as Maintenance (or similar)
+#define IS_MAINTENANCE_AREA (1<<16)
```
```suggestion
```

OrionTheFox commented on code/_globalvars/bitfields.dm:
```diff
@@ -67,6 +67,7 @@ DEFINE_BITFIELD(area_flags, list(
 	"NO_ALERTS" = NO_ALERTS,
 	"AREA_USES_STARLIGHT" = AREA_USES_STARLIGHT,
 	"PERSISTENT_ENGRAVINGS" = PERSISTENT_ENGRAVINGS,
+	"IS_MAINTENANCE_AREA" = IS_MAINTENANCE_AREA,
```
```suggestion
```

OrionTheFox commented on code/game/area/space_station_13_areas.dm:
```diff
@@ -168,7 +168,7 @@ NOTE: there are two lists of areas in the end of this file: centcom and station
 	name = "Generic Maintenance"
 	ambience_index = AMBIENCE_MAINT
-	area_flags = BLOBS_ALLOWED | UNIQUE_AREA | CULT_PERMITTED | PERSISTENT_ENGRAVINGS
+	area_flags = BLOBS_ALLOWED | UNIQUE_AREA | CULT_PERMITTED | PERSISTENT_ENGRAVINGS | IS_MAINTENANCE_AREA
```
```suggestion
	area_flags = BLOBS_ALLOWED | UNIQUE_AREA | CULT_PERMITTED | PERSISTENT_ENGRAVINGS
```

OrionTheFox commented on code/modules/mob/living/simple_animal/bot/floorbot.dm:
```diff
@@ -368,12 +370,20 @@
+				F.broken = FALSE
+				F.burnt = FALSE
+				F.icon_state = initial(F.icon_state)
+			else if(is_this_maints.area_flags & IS_MAINTENANCE_AREA)	//place catwalk if it's plating and we're in maints
```
```suggestion
			else if(istype(is_this_maints, /area/maintenance))	//place catwalk if it's plating and we're in maints
```

Commit:
Scrapping the Bitflag idea

If we need a bitflag, it shouldnt be for an area thats meant to be a base-type anyways

OrionTheFox commented:
okay
one FINAL test. Hopefully.

OrionTheFox commented:
> okay one FINAL test. Hopefully.

Test successful!

LemonInTheDark commented:
Let's jive

NanoCats commented:
I just want to mention that I woke up today to discover that this tread got Github to e-mail me a few dozen times.

LemonInTheDark commented:
cucked

OrionTheFox commented:
@LemonInTheDark I wanna thank you for helping me thru the thorough reviews; even the bitfield stuff taught me new things, so in the end I feel like I learned some real good code stuffs :) so, again, thank you

LemonInTheDark commented:
It's always fun to teach someone who's willing to learn, I'm glad it was helpful to you. look forward to what you do in future
